lms/djangoapps/monitoring/tests/test_middleware.py

```diff
@@ -0,0 +1,113 @@
+        self._assert_code_owner_custom_metrics(
+            mock_view_func, mock_set_custom_metric, expected_code_owner=expected_owner
+        )
+
+    @patch('lms.djangoapps.monitoring.middleware.set_custom_metric')
+    def test_code_owner_no_mappings(self, mock_set_custom_metric):
+        mock_view_func = self._create_view_func_mock('xblock')
+        self.middleware.process_view(None, mock_view_func, None, None)
+        mock_set_custom_metric.assert_not_called()
+
+    @override_settings(CODE_OWNER_MAPPINGS=['invalid_mapping_list'])
+    @patch('lms.djangoapps.monitoring.middleware.set_custom_metric')
+    def test_code_owner_mapping_error(self, mock_set_custom_metric):
+        mock_view_func = self._create_view_func_mock('xblock')
+        self.middleware.process_view(None, mock_view_func, None, None)
+        self._assert_code_owner_custom_metrics(mock_view_func, mock_set_custom_metric, has_error=True)
+
+    @patch('lms.djangoapps.monitoring.middleware.set_custom_metric')
+    @patch('lms.djangoapps.monitoring.middleware._get_view_func_module', side_effect=_mock_get_view_func_module)
+    def test_mapping_performance(self, mock_get_view_func_module, mock_set_custom_metric):
+        code_owner_mappings = []
+        # create a long list of mappings that are nearly identical
+        for n in range(1, 200):
+            path = 'openedx.core.djangoapps.{}'.format(n)
+            code_owner_mappings.append([path, 'team-red'])
+        # test a module that matches nearly to the end, but doesn't actually match
+        mock_view_func = self._create_view_func_mock('openedx.core.djangoapps.XXX.views')
+        with override_settings(CODE_OWNER_MAPPINGS=code_owner_mappings):
+            time = timeit.timeit(lambda: self.middleware.process_view(None, mock_view_func, None, None), number=100)
+        average_time = time / 100
+        self.assertTrue(average_time < 0.005, 'Mapping takes {}s which is too slow.'.format(average_time))
+        self._assert_code_owner_custom_metrics(mock_view_func, mock_set_custom_metric)
+
+    def _assert_code_owner_custom_metrics(
+        self, view_func, mock_set_custom_metric, expected_code_owner=None, has_error=False
+    ):
+        # default of unknown is always called in the current implementation, but this isn't required
+        self.assertEqual(mock_set_custom_metric.mock_calls[0], call('view_func_module', view_func.__module__))
+        if expected_code_owner:
+            self.assertEqual(mock_set_custom_metric.mock_calls[1], call('code_owner', expected_code_owner))
+        if has_error:
+            self.assertEqual(mock_set_custom_metric.mock_calls[1].args[0], 'code_owner_mapping_error')
+        time_call_index = 2 if expected_code_owner or has_error else 1
+        self.assertEqual(mock_set_custom_metric.mock_calls[time_call_index].args[0], 'code_owner_mapping_time')
+        self.assertTrue(mock_set_custom_metric.mock_calls[time_call_index].args[1] < 0.005)
```

Comment:
**Reviewer Note:** This is sort of a worst-case-scenario test.  It started failing at 0.001.  Maybe we should document what we think is actually acceptable before we decide we must improve performance.  Or, do we want to preemptively improve it?

Reply:
Optimizations complete.